core/src/main/java/com/graphhopper/util/details/PathDetails.java

```diff
@@ -0,0 +1,135 @@
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Holds the details for a path
+ *
+ * @author Robin Boldt
+ */
+public class PathDetails {
+
+    private final String name;
+    private boolean isOpen = false;
+
+    private Detail currentDetail = new Detail();
+    private List<Detail> pathDetails = new ArrayList<>();
+
+    public PathDetails(String name) {
+        this.name = name;
+    }
+
+    /**
+     * It is only possible to open one interval at a time.
+     *
+     * @param value
+     */
+    public void startInterval(Object value) {
+        if (isOpen) {
+            throw new IllegalStateException("Path details is already open with value: " + this.currentDetail.value + " trying to open a new one with value: " + value);
+        }
+        this.currentDetail = new Detail(value);
+        isOpen = true;
+    }
+
+    /**
+     * Ending intervals multiple times is safe, we only write the interval if it was opened.
+     * <p>
+     * Writes the interval to the pathDetails
+     *
+     * @param numberOfPoints Length of the PathDetail
+     */
+    public void endInterval(int numberOfPoints) {
+        // We don't want PathDetails
+        if (isOpen && numberOfPoints > 0) {
+            this.currentDetail.numberOfPoints = numberOfPoints;
+            pathDetails.add(this.currentDetail);
+        }
+        isOpen = false;
+    }
+
+    @JsonIgnore
+    public List<Detail> getDetails() {
+        return pathDetails;
+    }
+
+    public void merge(PathDetails pD) {
+        if (!this.name.equals(pD.getName())) {
+            throw new IllegalArgumentException("Only PathDetails with the same name can be merged");
+        }
+        List<Detail> otherDetails = pD.getDetails();
+
+        // Make sure that pathdetails are merged correctly at waypoints
+        if(!this.pathDetails.isEmpty() && !otherDetails.isEmpty()){
+            Detail lastDetail = pathDetails.get(pathDetails.size()-1);
+            // Add Via Point
+            lastDetail.numberOfPoints++;
+            if(lastDetail.value.equals(otherDetails.get(0).value)){
+                lastDetail.numberOfPoints += otherDetails.get(0).numberOfPoints;
+                otherDetails.remove(0);
+            }
+        }
+
+        this.pathDetails.addAll(otherDetails);
+    }
+
+    public String getName() {
+        return this.name;
+    }
+
+    @JsonProperty("details")
+    public Map<Object, List<int[]>> getPathDetailsMap() {
+        Map<Object, List<int[]>> detailsMap = new HashMap<>();
+
+        int pointer = 0;
+
+        for (Detail detail : this.pathDetails) {
+            List<int[]> detailIntervals;
+            if (detailsMap.containsKey(detail.value)) {
+                detailIntervals = detailsMap.get(detail.value);
+            } else {
+                detailIntervals = new ArrayList<>();
+            }
+            detailIntervals.add(new int[]{pointer, pointer + detail.numberOfPoints});
+            detailsMap.put(detail.value, detailIntervals);
```

Comment:
Yes :). Putting an existing element should be O(1), so I didn't bother. Can change this though.